crates/sdk/tests/test-client/src/main.rs

```diff
@@ -0,0 +1,1212 @@
+use spacetimedb_sdk::{
+    identity::{identity, load_credentials, once_on_connect, save_credentials},
+    once_on_subscription_applied,
+    reducer::Status,
+    subscribe,
+    table::TableType,
+};
+
+#[allow(clippy::too_many_arguments)]
+#[allow(clippy::large_enum_variant)]
+#[rustfmt::skip]
+mod module_bindings;
+
+use module_bindings::*;
+
+mod test_counter;
+use test_counter::TestCounter;
+
+mod simple_test_table;
+use simple_test_table::insert_one;
+
+mod pk_test_table;
+use pk_test_table::insert_update_delete_one;
+
+mod unique_test_table;
+use unique_test_table::insert_then_delete_one;
+
+const LOCALHOST: &str = "http://localhost:3000";
+
+fn db_name_or_panic() -> String {
+    std::env::var("SPACETIME_SDK_TEST_DB_NAME").expect("Failed to read db name from env")
+}
+
+/// Register a panic hook which will exit the process whenever any thread panics.
+///
+/// This allows us to fail tests by panicking in callbacks.
+fn exit_on_panic() {
+    // The default panic hook is responsible for printing the panic message and backtrace to stderr.
+    // Grab a handle on it, and invoke it in our custom hook before exiting.
+    let default_hook = std::panic::take_hook();
+    std::panic::set_hook(Box::new(move |panic_info| {
+        // Print panic information
+        default_hook(panic_info);
+
+        // Close the websocket gracefully before exiting.
+        spacetimedb_sdk::disconnect();
+
+        // Exit the process with a non-zero code to denote failure.
+        std::process::exit(1);
+    }));
+}
+
+fn main() {
+    env_logger::init();
+    exit_on_panic();
+
+    let test = std::env::args()
+        .nth(1)
+        .expect("Pass a test name as a command-line argument to the test client");
+
+    match &*test {
```

Comment:
It might make it more managable to break these tests out into separate files, but also fine to keep as a monolith if you feel its preferable 👍

Reply:
I don't think we'd win much, if anything, from breaking the tests into separate files within a single `test-client` crate. I think it would be ideal to break the tests into a large number of module/client-project pairs (which is why I wrote the test harness to support that), but that would lead to quite a bit of crate proliferation in our workspace - in addition to 15 client crates and probably 12 or 13 modules, we'd need a shared utils crate for the modules, and another for the clients. If you think that's an improvement, I'm happy to do it (it would also help with parallelizing the tests more), but I wasn't sure about doing it of my own volition.